### ISSUE TITLE
[BUG] golang client mishandles return type of string with content-type = application/json

### ISSUE THREAD

**Issue by jpriebe** [OpenAPITools/openapi-generator#9525]
#### Bug Report Checklist

- [*] Have you provided a full/minimal spec to reproduce the issue?
- [*] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [*] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [*] Have you searched for related issues/PRs?
- [*] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

API responses of JSON-encoded simple string value with content-type = application/json are mishandled by the golang client and are returned to the caller without JSON-decoding.

##### openapi-generator version

latest (5.1.1)

##### OpenAPI declaration file content or url

https://gist.github.com/jpriebe/3bfc34a63b586616bc22190b84a5440e

##### Generation Details

```
docker run \
    --rm \
    --volume "$PWD:/workdir" \
    openapitools/openapi-generator-cli:latest \
    generate \
    --input-spec /workdir/openapi.yaml \
    --generator-name go \
    --package-name "client" \
    --output /workdir
```

##### Steps to reproduce

- build an openapi spec for an API that produces application/json and has an endpoint that returns a string
- implement a server based on the spec
- generate a golang client with the spec
- use the client to make a call to the endpoint in question
- assuming the server returned a JSON-encoded string (as it should), the client will return JSON-encoded string instead of the raw string of the response

##### Suggest a fix

I think the problem is here:

https://github.com/OpenAPITools/openapi-generator/blob/60dcf8613f6b75b606757c4a991fe018ad10ee99/modules/openapi-generator/src/main/resources/go/client.mustache#L418

When the caller expects a string, it will pass in a pointer to a string to the decode() function.  The login on line 418 will pop the HTTP response body directly into the string and return.

I think that the decode() function should be modified in one of the following ways:

- add a check for content-type = text/plain to the logic block at line 418
- move the logic block at line 418 to line 442

Bottom line: if the server's content-type is application/json, the decode() function should be decoding JSON no matter what the expected go type is.

**Comment by jpriebe** [comment]
As an example of a response where this happens:

```
HTTP/1.1 200 OK
Content-Length: 10
Connection: keep-alive
Content-Security-Policy: default-src 'self'
Content-Type: application/json
Date: Wed, 19 May 2021 14:11:47 GMT
Server: nginx/1.15.2
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block

"foo"

```

the return value from the golang client call to this endpoint should be "foo"; instead, it is "\"foo\"\n".

**Comment by jpriebe** [comment]
One more bit of information.  RFC 8259, the most current specification for JSON, does explicitly say that a JSON text can be a string (older specs said that a JSON text had to be an object or an array)

https://datatracker.ietf.org/doc/html/rfc8259

> A JSON text is a serialized value.  Note that certain previous
> specifications of JSON constrained a JSON text to be an object or an
> array.  Implementations that generate only objects or arrays where a
> JSON text is called for will be interoperable in the sense that all
> implementations will accept these as conforming JSON texts.

**Comment by wtrocki** [comment]
Note that the same happens when os.file is used. Decode function should check target of decoding along with content types.

**Comment by wtrocki** [comment]
@jpriebe chances of this being picked are extremely low. Having PR will help
